### PR TITLE
Attempt not to 429 burst submissions if a delay will do

### PR DESCRIPTION
**Pull request by dannyroberts**
##### SUMMARY
When receiving a request that would be throttled by the second or
minute, wait and retry for up to 15 seconds. This is less load on the
server than rehandling client retries, and a better experience for the
user. This is especially true in the case of API usage, where the script
may not handle 429s well.

This also collects a metric on these throttle delays, and changes the test throttle mode (which we currently don't use) to match this behavior.

Only affects SaaS.

### CODE REVIEW

**Review by millerdev — APPROVED** [review]

**Review by biyeun — APPROVED** [review]

File: corehq/apps/receiverwrapper/rate_limiter.py
Comment (millerdev) [inline corehq/apps/receiverwrapper/rate_limiter.py]: `('hour', 'day', 'week')` is the default, right?

https://github.com/dimagi/commcare-hq/blob/1c1679314629385ed7dd7fc684856a4e14d6a0e1/corehq/project_limits/rate_limiter.py#L73

I'm curious why you're passing it here?

FWIW, I didn't really understand why you were originally passing `('second', 'minute')` before https://github.com/dimagi/commcare-hq/commit/1c1679314629385ed7dd7fc684856a4e14d6a0e1, but I glossed over it in my review thinking you had some good reason that I was not understanding. Is this parameter ever used to pass anything other than the default? If not, maybe it should be removed? In my view extra parameters like this make the code more complicated and harder to reason about.

Comment (dannyroberts) [reply corehq/apps/receiverwrapper/rate_limiter.py]: I went ahead and removed the explicit kwarg in this call, because it was passing in the default value, and I agree that's redundant/confusing.

I had originally in uncommitted code called this kwarg `windows_to_wait_on` with the inverted meaning, and when I changed it I forgot to change `('second', 'minute')` in this call to `('hour', 'day', 'week')` (which is an example of how passing in a kwarg you don't need to is can be error prone).

Semantically, this kwarg (`windows_not_to_wait_on`) means "If the usage has it its limits for any of the following windows, reject it quickly and don't bother waiting". The idea is that if you've already used your hour's allotment, it's very unlikely that waiting 15 seconds is going to get you under the limit; whereas if you're over your second-window allotment, you may very well just be able to wait one second and be under the limit again. (The minute window is somewhere in between, and I decided to wait on it as well.)

**Comment by dannyroberts** [comment]
OK I tested this on staging and it appears to be doing what I want it to be doing, including the metrics

**Comment by dannyroberts** [comment]
@biyeun @millerdev I'll be offline for the afternoon, but feel free to merge this once you approve, in case there's a deploy going out today

File: corehq/apps/receiverwrapper/rate_limiter.py
Comment (biyeun) [inline corehq/apps/receiverwrapper/rate_limiter.py]: absolutely not blocking. but I recently learned about `f` strings from @millerdev 
`f'duration:{duration_tag}'` might be more readable? 😄

**Review by biyeun — APPROVED** [review]

File: corehq/apps/receiverwrapper/rate_limiter.py
Comment (dannyroberts) [reply corehq/apps/receiverwrapper/rate_limiter.py]: Yeah they're new to me too! https://github.com/dimagi/commcare-hq/pull/26507

**Pull request merged**
